docs: add etcd quroum recovery (#1696)
canonical/k8s-snap: docs/canonicalk8s/snap/howto/restore-quorum.md

# How to recover a cluster after quorum loss

> Highly available {{product}} clusters can survive losing one or more

Author: Done

nodes. Both [etcd] and [Dqlite] implement a [Raft] based protocol
where an elected leader holds the definitive copy of the database, which is
then replicated on two or more secondary nodes.

> When the a majority of the nodes are lost, the cluster becomes unavailable.

Author: I'm a bit reluctant to replicate the upstream docs here, specially now that we have a reference to it. WDYT?

If at least one database node survived, the cluster can be recovered using the
steps outlined in this document.

```{note}
This guide can be used to recover the {{product}} managed datastore,
which can be either etcd or Dqlite. Persistent volumes on the lost nodes are 
*not* recovered.
```

If you have set Dqlite as the datastore, please consult the 
[Dqlite configuration reference] before moving forward.

## etcd

### Take an etcd snapshot

Choose one of the remaining cluster nodes that has the most recent
version of the Raft log.

Install `etcdctl` and `etcdutl` binaries following the 
[etcd upstream installation instructions]. With sudo privilege, 
run `etcdctl` to verify you have access to etcd cluster node's data:

```
sudo etcdctl --cacert /etc/kubernetes/pki/etcd/ca.crt \
        --cert /etc/kubernetes/pki/apiserver-etcd-client.crt \
        --key /etc/kubernetes/pki/apiserver-etcd-client.key \
        snapshot save snapshot.db
```

> Follow the [upstream instructions] to take an snapshot of the Keyspace. 

Author: Done, thanks a lot

### Stop {{product}} services on all nodes

Before recovering the cluster, all remaining {{product}} services
must be stopped on every node:

```
sudo snap stop k8s
```

> ### Recover the k8sd database

Author: Done, please let me know if there is a need for more elaboration

> ### Recover the k8sd database

Author: This is replicated in both etcd and dqlite sections since the steps are kind of tangled like
```
for_all_nodes(stop service ---> k8sd recovery ---> datastore-specific recovery ----> start services)
```
I don't know if it works correctly if we change it to:
```
for_all_nodes(stop services ---> k8sd recovery ----> start services]) ---> for_all_nodes(datastore-specific recovery)
```
But it would enable us to take the k8sd part out and explain it once regardless of k8s datastore. WDYT?

Choose one of the remaining healthy cluster nodes that has the most recent

> version of the Raft log. Use the `cluster-recover` command to reconfigure 

Author: Should we check the k8sd raft log or the k8s datastore log?

the Raft members and generate recovery tarballs that are used to restore the 
cluster datastore on lost nodes. The command is an interactive tool that 
allows you modify the relevant files and provides useful hints at each step.

> 

Author: Done

```note
By default, `cluster-recover` will recover both Dqlite cluster and Kubernetes 
datastores. When etcd is used, the ``--skip-k8s-dqlite`` flag is needed to 
instruct `cluster-recover` to ignore the Dqlite Kubernetes datastore.
```

Adjust the log level for additional debug messages by increasing its
value. Database backups are created before making any changes.

Copy the generated ``recovery_db.tar.gz`` to all remaining nodes at
``/var/snap/k8s/common/var/lib/k8sd/state/recovery_db.tar.gz``. When the k8sd
service starts, it will load the archive and perform the necessary recovery
steps.

### Restore the etcd snapshot

Run the following command on all remainining live nodes to reconfigure the 
etcd membership:

```
etcdutl snapshot restore snapshot.db \
      --name=<NAME> \
      --initial-advertise-peer-urls=<ADVERTISE_PEER_URLS> \
      --initial-cluster=<INITIAL_CLUSTER> \
      --data-dir /var/snap/k8s/common/var/lib/etcd/data
```

Get the `<NAME>` and `<ADVERTISE_PEER_URLS>` for each node by 
executing the following command:

```
args=$(grep -E '^--(name|initial-advertise-peer-urls)=' /var/snap/k8s/common/args/etcd | xargs)
```

> 

Author: Done

The `<INITIAL_CLUSTER>` will be the comma-separated list of all remaining 
cluster members fetched from each node by running:

```

> args=$(grep -E '^--(initial-cluster)=' /var/snap/k8s/common/args/etcd | xargs)

Author: The first one is per node and the second regarding `initial-cluster` should be gathered from all the nodes and manually curated. I felt separating it would make it more explicit. Does that make sense? We should have examples as you mentioned though

> args=$(grep -E '^--(initial-cluster)=' /var/snap/k8s/common/args/etcd | xargs)

Reviewer: It is not so clear what you mean about "the initial-cluster should be gathered from all the nodes and manually curated". Could you please explain?

```

### Start the services 

> For each node, start the {{product}} services by running:

Reviewer: Do you think we should tell users that the order doesn't matter? Is that something they will wonder about?

# How to recover a cluster after quorum loss

Highly available {{product}} clusters are designed to tolerate the loss of 
one or more nodes. Both [etcd] and [Dqlite] use [Raft] protocol
where an elected leader holds the definitive copy of the database, which is
then replicated on two or more secondary nodes.

When the a majority of the nodes are lost, the cluster becomes unavailable.
If at least one database node survived, the cluster can be recovered using the
steps outlined in this document.

> ```{warning}

Author: Done

This guide can be used to recover the {{product}} managed datastore,
which can be either etcd or Dqlite. Persistent volumes on the lost nodes are 
*not* recovered.
```

{{product}} relies on two separate distributed datastores. The first 
is a Dqlite-based cluster datastore used to manage the distribution's state. 
The second database stores the Kubernetes objects' state and is either etcd 
by default or Dqlite, depending on user configuration. For more information, 
please see [the architecture guide].

> Let's suppose you had a cluster with 7 control plane nodes, and 4 of them lost 

Author: Luise had in mind that an overview could help the user. I changed those two paragraphs to bullet points covering the steps. Please let me know if it needs further removal/adjustments

> Let's suppose you had a cluster with 7 control plane nodes, and 4 of them lost 

Reviewer 2: <img width="1415" height="343" alt="image" src="https://github.com/user-attachments/assets/5d3c08b6-23c5-4281-b730-3d7bb6fc08cf" />
So generally we try to give the user an overview of _what_ will be achieved in the guide at the beginning, not necessarily a step by step of _how_ it will be achieved as they can scroll down to find out more. Additionally, the floating navigation on the right also allows the user to see the steps they will be following. I agree with Louise that we were probably missing a bit of context initially but I believe you nailed it in the first three paragraphs and therefore this one is not required. You successfully tell the user what you are going to do (that because of HA we can be recover), the requirements for recovering (at least one has to have survived to recover) and which databases are recovered in the guide.

connection to the cluster. At this point, the cluster cannot form a quorum and 
ceases to operate. To bring the cluster back online with the remaining 3 
nodes, we are going to start by stopping the services on all machines. Then, 
pick one healthy machine to guide the recovery. The recovery process includes 
reconfiguring the cluster membership on both {{product}} datastores to remove 
the 4 lost nodes, instructing the datastores to form a quorum with the 
remainin 3 nodes, and replicating the most recent Raft log so that all the 
remaining nodes have the same view of the distributed state. There’s a helpful 
tool that will walk you through some of the process. It will create the 
files needed to restore the rest of the machines.

> Whether your setup uses etcd or Dqlite, just follow the instructions to

Reviewer 2: I would also remove this paragraph

restore the data safely. Once that’s done, restart the services on each
machine. The system will reconnect, and before long, your cluster will be up
and running again.

If you have set up Dqlite as the datastore, please consult the 
[Dqlite configuration reference] before moving forward.

## Stop {{product}} services on all nodes

Before recovering the cluster, all remaining {{product}} services
must be stopped on every node:

```
sudo snap stop k8s
```

## Recover the cluster datastore 

Choose one of the remaining healthy cluster nodes that has the most recent
version of the Raft log. 

> ```{note}

Author: Done

To find the node with the most recent log entries, navigate to

> /var/snap/k8s/common/var/lib/k8sd/state/database on each node and

Author: Done

look at all the segment files matching the format 
`0000000000006145-0000000000006823` (this is just an example).
For each node, identify the segment file with the highest 
end-segment index (e.g., 6823 in the example), then compare across 
nodes and select the one with the highest index overall.
```

Use the `cluster-recover` command to reconfigure 
the Raft members and generate recovery tarballs that are used to restore the 
cluster datastore on lost nodes. The command is an interactive tool that 
allows you to modify the relevant files and provides useful hints at each step.
On the node with the most recent Raft logs, run:

`````{tabs}

````{group-tab} etcd

```
sudo /snap/k8s/current/bin/k8sd cluster-recover \
    --state-dir=/var/snap/k8s/common/var/lib/k8sd/state \
    --k8s-dqlite-state-dir=/var/snap/k8s/common/var/lib/k8s-dqlite \
    --log-level 0
    --skip-k8s-dqlite
```

````

````{group-tab} Dqlite

```
sudo /snap/k8s/current/bin/k8sd cluster-recover \
    --state-dir=/var/snap/k8s/common/var/lib/k8sd/state \
    --k8s-dqlite-state-dir=/var/snap/k8s/common/var/lib/k8s-dqlite \
    --log-level 0
```

````

`````

```{note}

> By default, the command will recover both Dqlite databases. If one of the

Reviewer 2: ```suggestion
By default, the command will recover both the cluster and Kubernetes datastore Dqlite databases. If one of the
```

databases needs to be skipped, use the ``--skip-k8sd`` or ``--skip-k8s-dqlite``
flags. This can also be useful when using an external etcd database.
```

Use the command to update the ``cluster.yaml`` file, changing the role of the 
lost nodes to "spare". Additionally, verify the addresses and IDs specified 
in ``cluster.yaml``, ``info.yaml`` and ``daemon.yaml`` are correct, 
especially if database files were moved across nodes.

> The above command will reconfigure the Raft members and create recovery

Reviewer 2: I think this is duplicated and can be removed

# How to recover a cluster after quorum loss

> Highly available {{product}} clusters are designed to tolerate losing 

Reviewer: ```suggestion
Highly available {{product}} clusters are designed to tolerate the loss of
```

nodes. Both [etcd] and [Dqlite] implement a [Raft] based protocol
where an elected leader holds the definitive copy of the database, which is
then replicated on two or more secondary nodes.

> 

Author: Done

steps outlined in this document.

```{note}
This guide can be used to recover the {{product}} managed datastore,
which can be either etcd or Dqlite. Persistent volumes on the lost nodes are 
*not* recovered.
```

If you have set Dqlite as the datastore, please consult the 
[Dqlite configuration reference] before moving forward.

> # etcd

Author: Done

## Take an etcd snapshot

Choose one of the remaining alive cluster nodes that has the most recent
version of the Raft log.

Install `etcdctl` and `etcdutl` binaries following the 

> [etcd upstream installation instructions]. Run this command with sudo 

Author: Done

privilege to verify you have access to etcd cluster node's data:

```
sudo etcdctl --cacert /etc/kubernetes/pki/etcd/ca.crt \
        --cert /etc/kubernetes/pki/apiserver-etcd-client.crt \
        --key /etc/kubernetes/pki/apiserver-etcd-client.key \
        snapshot save snapshot.db
```

Follow the [upstream instructions] to take an snapshot of the keyspace. 

## Stop {{product}} services on all nodes

Before recovering the cluster, all remaining {{product}} services

> must be stopped. Use the following command on every node:

Author: Done

```
sudo snap stop k8s
```

## Recover the k8sd database

> Update the ``cluster.yaml`` file, changing the role of the lost nodes to

Author: Done

> "spare". Additionally, double check the addresses and IDs specified in

Author: Done

``cluster.yaml``, ``info.yaml`` and ``daemon.yaml``, especially if database
files were moved across nodes.

> The following command guides us through the recovery process, prompting a text

Author: I agree that this section has been a bit vague even for dqlite. When you run the `cluster-recover` command, it prompts `nano` and lets the user to modify three files, and for each file there is a comment section providing hints on how to modify the files. The editor is referring to this feature of the tool. I have rephrased it to try and make it a more clear using your suggestions. Please let me know how do you feel about it

editor with informative inline comments for each of the Dqlite configuration
files.

```
sudo /snap/k8s/current/bin/k8sd cluster-recover \
    --state-dir=/var/snap/k8s/common/var/lib/k8sd/state \
    --k8s-dqlite-state-dir=/var/snap/k8s/common/var/lib/k8s-dqlite \
    --log-level 0
    --skip-k8s-dqlite

> ```

Author: Done

Please adjust the log level for additional debug messages by increasing its
value. The command creates database backups before making any changes.

Copy the generated ``recovery_db.tar.gz`` to all remaining nodes at
``/var/snap/k8s/common/var/lib/k8sd/state/recovery_db.tar.gz``. When the k8sd
service starts, it will load the archive and perform the necessary recovery
steps.

## Restore the etcd snapshot

> Run the following command on all remanining alive nodes to reconfigure the 

Author: Done

steps outlined in this document.

> ```{note}

Reviewer: Let's make this a warning please.

This guide can be used to recover the {{product}} managed datastore,
which can be either etcd or Dqlite. Persistent volumes on the lost nodes are 
*not* recovered.
```

> {{product}} smoothlyrelies on two separate distributed datastores. The first 

Reviewer: ```suggestion
{{product}} smoothly relies on two separate distributed datastores. The first 
```

Author: It was a typo :D smoothly should be removed, thanks a lot

> is a Dqlite-based cluster datastore used to manage the {{product}} itself. 

Reviewer: ```suggestion
is a Dqlite-based cluster datastore used to manage the distribution's state. 
```

Author: Done

> The second is the Kubernetes backend datastore, which stores 
> Kubernetes objects' state and can be either etcd (default) or Dqlite, depending 
> on user configuration. For more information, please see [the architecture].

Reviewer: ```suggestion
The second database stores the
Kubernetes objects' state and is either etcd by default or Dqlite, depending 
on user configuration. For more information, please see [the architecture guide].
```

Author: Done

> Let's suppose you had a cluster with 7 control plane nodes, and 4 of them lost

Reviewer: Most clusters use three nodes a few may use 5, I've never heard of anyone running a 7 node cluster (not to say they don't exist) but let's use the most common scenario: 3 nodes.

connection to the cluster. At this point, the cluster cannot form a quorum and
ceases to operate. To bring the cluster back online with the remaining 3 nodes,
we are going to start by stopping the services on all machines. Then, pick one
healthy machine to guide the recovery. The recovery process includes
reconfiguring the cluster membership on both {{product}} datastores to remove
the 4 lost nodes, instructing the datastores to form a quorum with the 
remainin 3 nodes, and replicating the most recent Raft log so that all the
remaining nodes have the same view of the distributed state. There’s a helpful
tool that will walk you through some of the process. It will create the

> files needed to restore the rest of the machines.

Reviewer: Could this be a quick bullet point overview, bonus points if it links to the section headers.

Author: Tried a few combinations but didn't come out clean. They became something like:
- stop services
- do everything in between
- start the services

Whether your setup uses etcd or Dqlite, just follow the instructions to
restore the data safely. Once that’s done, restart the services on each
machine. The system will reconnect, and before long, your cluster will be up
and running again.

> If you have set Dqlite as the datastore, please consult the 

Reviewer: ```suggestion
If you have set up Dqlite as the datastore, please consult the 
```

[Dqlite configuration reference] before moving forward.

## Stop {{product}} services on all nodes

Before recovering the cluster, all remaining {{product}} services
must be stopped on every node:

```
sudo snap stop k8s
```

## Recover the cluster datastore 

> Choose one of the remaining healthy cluster nodes that has the most recent

Reviewer: ```suggestion
Choose the cluster node with the most recent
```
How does this work in etcd?

    --skip-k8s-dqlite
```

````

````{group-tab} Dqlite

```
sudo /snap/k8s/current/bin/k8sd cluster-recover \
    --state-dir=/var/snap/k8s/common/var/lib/k8sd/state \
    --k8s-dqlite-state-dir=/var/snap/k8s/common/var/lib/k8s-dqlite \
    --log-level 0
```

````

`````

```{note}
By default, the command will recover both Dqlite databases. If one of the
databases needs to be skipped, use the ``--skip-k8sd`` or ``--skip-k8s-dqlite``
flags. This can also be useful when using an external etcd database.
```

Use the command to update the ``cluster.yaml`` file, changing the role of the 
lost nodes to "spare". Additionally, verify the addresses and IDs specified 
in ``cluster.yaml``, ``info.yaml`` and ``daemon.yaml`` are correct, 
especially if database files were moved across nodes.

The above command will reconfigure the Raft members and create recovery
tarballs that are used to restore the lost nodes, once the Kubernetes 
datastore configuration is updated.

Adjust the log level for additional debug messages by increasing its
value. Database backups are created by the command before making any changes.

Copy the generated ``recovery_db.tar.gz`` to all remaining nodes at

> ``/var/snap/k8s/common/var/lib/k8sd/state/recovery_db.tar.gz``. When the k8sd
> service starts, it will load the archive and perform the necessary recovery

Reviewer: ```suggestion
``/var/snap/k8s/common/var/lib/k8sd/state/recovery_db.tar.gz``. When we restart k8sd
in a later step, it will load the archive and perform the necessary recovery
```
Let's make it clear that the user should not restart k8sd here.

Author: Done, thank you

steps.

> ```{note}

Reviewer: this should be closer to the command? Maybe combined with the earlier skip k8s dqlite note?

This allows automating the recovery procedure.
```

## Recover the Kubernetes datastore

`````{tabs}

````{group-tab} etcd

### Take an etcd snapshot

> Choose one of the remaining cluster nodes that has the most recent

Reviewer: How do I find the one with the most recent logs?

Reviewer 2: If it is the same as Dqlite, I would point to the previous step and ask them to redo it here. Otherwise, remove the "mos recent" requirement from this sentence

version of the Raft log.

Install `etcdctl` and `etcdutl` binaries following the 
[etcd upstream installation instructions]. With sudo privilege, 
run `etcdctl` to verify you have access to etcd cluster node's data:

```
sudo etcdctl --cacert /etc/kubernetes/pki/etcd/ca.crt \
        --cert /etc/kubernetes/pki/apiserver-etcd-client.crt \
        --key /etc/kubernetes/pki/apiserver-etcd-client.key \
        snapshot save snapshot.db
```

> Follow the [upstream instructions] to take a snapshot of the Keyspace. 

Reviewer: ```suggestion
For further information, review the [upstream instructions] as we follow their steps to take a snapshot of the Keyspace. 
```
Otherwise it sounds like we have to do some extra steps that are not here.

### Restore the etcd snapshot

Run the following command on all remainining live nodes to reconfigure the 
etcd membership:

```
etcdutl snapshot restore snapshot.db \
      --name=<NAME> \
      --initial-advertise-peer-urls=<ADVERTISE_PEER_URLS> \
      --initial-cluster=<INITIAL_CLUSTER> \
      --data-dir /var/snap/k8s/common/var/lib/etcd/data
```

Get the `<NAME>` and `<ADVERTISE_PEER_URLS>` for each node by 
executing the following command:

```
echo $(grep -E '^--(name|initial-advertise-peer-urls)=' /var/snap/k8s/common/args/etcd | xargs)
```

for each node the output could be something like:

```
--initial-advertise-peer-urls=https://10.246.154.125:2380 --name=node-1
```

The `<INITIAL_CLUSTER>` will be the comma-separated list of all remaining 
cluster members. The list should be created by gathering the results from
running the following command on each node:

```
echo $(grep -E '^--(initial-cluster)=' /var/snap/k8s/common/args/etcd | xargs)
```

Aggregate the results from all remaining nodes to form the `<INITIAL_CLUSTER>`
that should have a format like below:

```
node-1=https://10.246.154.125:2380,node-2=https://10.246.154.126:2380,node-3=https://10.246.154.127:2380

> ```

Reviewer: Can we please improve the UX? Are you sure we can't use one command and print each on a new line?

Author: For the first two options that would be easy. But the `initial-cluster` output from each node is different. For example in a two node cluster the first node would return:
```
--initial-cluster=juju-5cb100-16=https://10.246.154.36:2380
```
While the second joined node returns:
```
--initial-cluster=juju-5cb100-16=https://10.246.154.36:2380,juju-5cb100-17=https://10.246.154.191:2380
```

Maybe removing the command and trusting the user to figure that out could be a better option?

> 

Reviewer: Let's put it all together. Our example results in the following command:

  /var/snap/k8s/common/var/lib/k8s-dqlite
```

````

`````

## Start the services 

For each node, start the {{product}} services by running:

```
sudo snap start k8s
```

> Ensure that the services started successfully by using

Reviewer: ```suggestion
Confirm that the snap services started successfully by running
```

``sudo snap services k8s``. Use ``sudo k8s status --wait-ready`` to wait for the
cluster to become ready.

> Once a quorum is achieved the cluster will be reported as "highly available":

Reviewer: ```suggestion
Once our nodes are up, the cluster will be reported as "highly available" again:
```

Author: Done, thank you

> Once a quorum is achieved the cluster will be reported as "highly available":

Reviewer: Have we tested restoring the snapshots on 3 fresh nodes? Maybe we would have to perform some additional clean-up.
We can create another card for that, it's out of scope for this PR. Wdyt?

Author: That would be a good idea, I have not tested with adding new nodes
